File: src/priv/bajdcc/syntax/token/Token.java
package priv.bajdcc.syntax.token;

import priv.bajdcc.utility.Position;

/**
 * 单词
 * 
 * @author bajdcc
 */
public class Token {

	/**
	 * 单词类型
	 */
	public TokenType kToken = TokenType.ERROR;

	/**
	 * 数据
	 */
	public Object object = null;

	/**
	 * 位置
	 */
	public Position position = new Position();

	@Override
	public String toString() {
		StringBuilder sb = new StringBuilder();
		sb.append(String.format("%04d,%03d: %s %s", position.iLine,
				position.iColumn, kToken.getName(),
				object == null ? "(null)" : object.toString()));
		return sb.toString();
	}

	public static Token transfer(priv.bajdcc.lexer.token.Token token) {
		Token tk = new Token();
		tk.object = token.object;
		tk.position = token.position;
		switch (token.kToken) {
		case COMMENT:
			tk.kToken = TokenType.COMMENT;
			break;
		case EOF:
			tk.kToken = TokenType.EOF;
			break;
		case ERROR:
			tk.kToken = TokenType.ERROR;
			break;
		case ID:
			tk.kToken = TokenType.NONTERMINAL;
			break;
		case MACRO:
			tk.kToken = TokenType.HANDLER;
			break;
		case OPERATOR:
			tk.kToken = TokenType.OPERATOR;
			break;
		case STRING:
			tk.kToken = TokenType.TERMINAL;
			break;
		case WHITESPACE:
			tk.kToken = TokenType.WHITSPACE;
			break;
		case INTEGER:
			tk.kToken = TokenType.STORAGE;
			break;
		default:
			break;
		}
		return tk;
	}
}

File: src/main/java/com/bajdcc/LALR1/syntax/token/Token.kt
package com.bajdcc.LALR1.syntax.token

import com.bajdcc.util.Position
import com.bajdcc.util.lexer.token.Token as LexerToken
import com.bajdcc.util.lexer.token.TokenType as LexerTokenType

/**
 * 单词
 *
 * @author bajdcc
 */
class Token {

    /**
     * 单词类型
     */
    var type = TokenType.ERROR

    /**
     * 数据
     */
    var obj: Any? = null

    /**
     * 位置
     */
    var position = Position()

    override fun toString(): String {
        return String.format("%04d,%03d: %s %s", position.line,
                position.column, type.desc,
                obj?.toString() ?: "(null)")
    }

    companion object {

        fun transfer(token: LexerToken): Token {
            val tk = Token()
            tk.obj = token.obj
            tk.position = token.position
            when (token.type) {
                LexerTokenType.COMMENT -> tk.type = TokenType.COMMENT
                LexerTokenType.EOF -> tk.type = TokenType.EOF
                LexerTokenType.ERROR -> tk.type = TokenType.ERROR
                LexerTokenType.ID -> tk.type = TokenType.NONTERMINAL
                LexerTokenType.MACRO -> tk.type = TokenType.HANDLER
                LexerTokenType.OPERATOR -> tk.type = TokenType.OPERATOR
                LexerTokenType.STRING -> tk.type = TokenType.TERMINAL
                LexerTokenType.WHITESPACE -> tk.type = TokenType.WHITESPACE
                LexerTokenType.INTEGER -> tk.type = TokenType.STORAGE
                LexerTokenType.RESERVE -> tk.type = TokenType.ACTION
                else -> {
                }
            }
            return tk
        }
    }
}